synapse/cortex.py

```diff
@@ -6284,6 +6313,491 @@ async def iterTagPropRows(self, layriden, tag, prop, form=None, stortype=None, s
+            try:
+                s_msgpack.en({key: valu})
+            except s_exc.NotMsgpackSafe as exc:
+                raise s_exc.BadArg(mesg=f'Vault data must be msgpack safe.') from None
+
+            data[key] = valu
+
+        return await self._push('vault:data:set', iden, key, valu, delete)
+
+    @s_nexus.Pusher.onPush('vault:data:set')
+    async def _setVaultData(self, iden, key, valu, delete):
+        vault = self.reqVault(iden)
+        data = vault.get('data')
+
+        bidn = s_common.uhex(iden)
+
+        if delete:
+            if key in data:
+                data.pop(key)
+        else:
+            data[key] = valu
+
+        self.slab.put(bidn, s_msgpack.en(vault), db=self.vaultsdb)
+        return data
+
+    def listVaults(self):
+        '''
+        List all vaults.
+
+        Args: None
+
+        Raises: None
+
+        Yields: tuples of vault info: (<iden>, <name>, <type>, <scope>).
+        '''
+        for vault in self._getVaults():
+            yield vault
+
+    async def setVaultPerm(self, viden, iden, level):
+        '''
+        Set vault permissions.
+        Args:
+            viden (str): The iden of the vault to edit.
+            iden (str): Iden of the user/role to add permissions for.
+            level (int): Easy perms level.
+
+        Raises:
+            synapse.exc.NoSuchIden: Vault with `iden` does not exist.
+
+        Returns: Updated vault.
+        '''
+        vault = self.reqVault(viden)
+
+        scope = 'users'
+        ident = self.auth.user(iden)
+        if ident is None:
+            scope = 'roles'
+            ident = self.auth.role(iden)
+            if ident is None:
+                raise s_exc.NoSuchIden(mesg=f'Iden {iden} is not a valid user or role.')
+
+        await self._setEasyPerm(vault, scope, ident.iden, level)
+        permissions = vault.get('permissions')
+        return await self._push(('vault:set'), viden, 'permissions', permissions)
+
+    async def renameVault(self, iden, name):
+        '''
+        Rename a vault.
+
+        Args:
+            iden (str): Iden of the vault to rename.
+            name (str): New vault name.
+
+        Raises:
+            synapse.exc.NoSuchIden: Vault with `iden` does not exist.
+            synapse.exc.DupName: Vault with `name` already exists.
+
+        Returns: Updated vault.
+        '''
+        if self.getVaultByName(name) is not None:
+            raise s_exc.DupName(mesg=f'Vault with name {name} already exists.')
+
+        vault = self.reqVault(iden)
+
+        oldname = vault.get('name')
+
+        if name == oldname:
+            return vault
+
+        vault['name'] = name
+
+        return await self._push(('vault:set'), iden, 'name', name)
+
+    @s_nexus.Pusher.onPush('vault:set')
+    async def _setVault(self, iden, key, valu):
+        if key not in ('name', 'permissions'):
+            raise s_exc.BadArg('Only vault names and permissions can be changed.')
+
+        vault = self.reqVault(iden)
```

Comment:
If we did the following we can ensure we're always serializing a valid Vault structure:
```

        vault = self.reqVault(iden)

        vault[key] = valu

        reqValidVault(vault)

        bidn = s_common.uhex(iden)

        if key == 'name':
            name = vault.get('name')
            self.slab.delete(name.encode(), db=self.vaultsbynamedb)
            self.slab.put(valu.encode(), bidn, db=self.vaultsbynamedb)

        self.slab.put(bidn, s_msgpack.en(vault), db=self.vaultsdb)
        return vault
```

Reply:
Thanks for the suggestion!